### DISCUSSION TITLE
Can Patroni make DB cluster with zero data loss & zero down time?

### DISCUSSION THREAD

**Discussion by ganjarsetia** [patroni/patroni#3197] [Q&A] [answered]
I want to build Database cluster that have Recovery Point Objective (RPO = 0 / zero data loss), Recovery Time Objective (RTO = 0 / zero down-time) and fast. After I'm searching, then I found Patroni and its support Citus to make it fast (really love it!). So I'm new here

Example case & expected result: when running `pg_restore` and having 200 million rows, in 5 seconds the coordinator leader is switching because of failovers/switchovers, the process is not halted/interrupted.

Is it possible with Patroni? can we do that with Citus enabled?

**Comment by CyberDem0n** [comment, marked as answer]
Zero RPO - yes (almost), if you enable synchronous replication. Almost - because there are details, which you can read about in https://github.com/patroni/patroni/issues/3194

Zero RTO - no. It is just physically impossible to detect failed primary so fast. But, 20s-30s downtime on failover is realistic.

Also, please keep in mind that RPO is not only about HA, but also backups. What would you do if someone dropped a table with very important data?
In case of Citus it gets even more complicated, because you will not have continues backups and mus rely on [citus_create_restore_point](https://docs.citusdata.com/en/v12.1/develop/api_udf.html#citus-create-restore-point) function to create consistent restore points across worker nodes.

**Reply by ganjarsetia** [reply]
Thank you for the reply by the Author!

When you said `Zero RPO - yes (almost)` is it 0.99999999999% ? :laughing: 

This helps clarify the practical limitations around achieving zero RPO and zero RTO with Patroni, especially with synchronous replication. It makes sense that even with HA, there’s always a risk of minor data loss due to the nuances in detection time and failover processes.
